### PR TITLE
PLG Cloud.gov

### LINKED ISSUE #3046: PLG deployed in Cloud.gov
**Description:**
Following #831 Prometheus has been selected as one of the logging servers. We need to deploy both Prometheus and Grafana in cloud.gov prod space.

**Acceptance Criteria:** 

_As an engineer I want to have Prometheus/Grafana to be available in Cloud.gov for system logs_

- [ ] Prometheus, Loki, and Grafana are deployed in cloud.gov
- [ ] Grafana is connected to Promethues and Loki and can query the logs from Loki
- [ ] Backend apps can push logs to Loki
- [ ] Prometheus can pull metrics from backend apps and postgres exporters
- [ ] Loki has persistent log storage via S3
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] Deploy PLG
- [x] Connect Grafana to Prometheus and Loki
- [x] Update Loki to store logs in S3
- [x] Update backend to deploy promtail as a sidecar
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**

- Study security remarks for deploying Prometheus and document your decisions 
- Follow Prometheus best practices

**Supporting Documentation:**

- [Getting started](https://prometheus.io/docs/prometheus/latest/getting_started/)
- [Monitoring linux server](https://prometheus.io/docs/guides/node-exporter/)

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Add manifests to deploy PLG as binaries via binary buildpack
- Add local vs deployed configurations for PLG apps
- Added routing based on dev environment for the interim
- Added deploy script for PLG and PG exporters
- Fixed local Nginx proxying to Grafana

Pull request closes #3046 

Considerations:
- We need to move/should we move all PLG apps into the production env?
- We need to actually calculate memory requirements for PLG. My assumption is Loki: 2-4GB, Prometheus: 2-4GB ([calculator](https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion/)), Grafana: 1GB, 3 PG exporters: 3 * 24MB, 6 Backend Promtails: 6 * 64MB. [1]
- ~We can/should hook Grafana up to our RDS to have a more persistent set of datasources, dashboards, etc...~
- Need to play around with cloud.gov provisioned dashboards. I couldn't get it to work so I just uploaded them manually.
- Need to re-evaluate getting promtail to run with frontend
- Need to update/add our Promtail pipelines to parse log messages better so we can filter our logs dashboard more effectively [3]
- Consider moving deployment of all apps to docker containers to make life with PLG and frontend/backend deployment a bit easier.
-  ~Need to verify logs in Loki are queryable from days, weeks, months, etc ago~
- While TDP lives in cloud.gov it might be worth it to see if we can setup a [log drain](https://cloud.gov/docs/deployment/logs/#how-to-automatically-copy-your-logs-elsewhere) through cloud.gov to loki. Then we should get all of the cloud.gov logs also.
- Consider adding metric exporting for [Elastic](https://github.com/prometheus-community/elasticsearch_exporter)
- Consider adding metric exporting for [machine stats (cpu, memory, etc)](https://github.com/prometheus/node_exporter) [5]
- Consider using [Grafana Alerting](https://grafana.com/docs/grafana/latest/alerting/) vs [AlertManager](https://prometheus.io/docs/alerting/latest/alertmanager/) [2]
- Prometheus does not have persistent storage. Need to integrate with something like [Mimir](https://github.com/grafana/mimir) to support S3 based metric storage. [4]

## How to Test
PLG is deployed in the dev environment for the moment. If you would like to browse [Grafana](https://grafana.app.cloud.gov/grafana) I have opened a public route to it for the interim until we decide where PLG is going to live. Reach out to me for username and password. Once you're logged in, feel free to browse the dashboards. Note, the `Logs` dashboard only has logging information as far back as 09/25/2024 at ~9:40am ET since that is when promtail had it's first successful exports.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Prometheus, Loki, and Grafana are deployed in cloud.gov
- [x] Grafana is connected to Promethues and Loki and can query the logs from Loki
- [x] Backend apps can push logs to Loki
- [x] Prometheus can pull metrics from backend apps and postgres exporters
- [x] Loki has persistent log storage via S3
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: scripts/deploy-backend.sh
Comment (elipe17) [inline scripts/deploy-backend.sh:118]: Do we want PLG to live in prod? If so, need to update networking.

**Comment by raftmsohani** [comment]
So that we all don't have to deploy to test this, could you possibly post screen shots from running command in steps (only for changes to deploy command)?

**Comment by elipe17** [comment]
> So that we all don't have to deploy to test this, could you possibly post screen shots from running command in steps (only for changes to deploy command)?

@raftmsohani the only thing you might want to deploy is the backend to see new logs being pushed to loki. Otherwise, everything is already deployed so you can just inspect with CF CLI or browse grafana.

**Review by jtimpe — APPROVED** [review]
Looking good!

File: tdrs-backend/plg/prometheus/manifest.yml
Comment (elipe17) [inline tdrs-backend/plg/prometheus/manifest.yml:14]: Look into cloud.gov equivalent of ebs storage since app storage is transient

**Review by andrew-jameson — APPROVED** [review]

File: scripts/deploy-backend.sh
Comment (andrew-jameson) [reply scripts/deploy-backend.sh:118]: Yes, to be tracked via #3222 per Office Hours 10/9

File: scripts/deploy-frontend.sh
Comment (andrew-jameson) [inline scripts/deploy-frontend.sh:16]: Is this no longer used, can this be deleted?

File: tdrs-backend/docker-compose.yml
Comment (andrew-jameson) [inline tdrs-backend/docker-compose.yml:99]: Could this be simplified to something like this?

`./plg/grafana/:/usr/share/grafana/`

File: tdrs-backend/plg/deploy.sh
Comment (andrew-jameson) [inline tdrs-backend/plg/deploy.sh:1]: Does this require any env variables to be loaded up if ran manually?

File: tdrs-backend/plg/grafana/dashboards/logs_dashboard.json
Comment (andrew-jameson) [inline tdrs-backend/plg/grafana/dashboards/logs_dashboard.json:28]: Is this intentionally hardcoded?

File: tdrs-backend/plg/grafana/dashboards/logs_dashboard.json
Comment (andrew-jameson) [inline tdrs-backend/plg/grafana/dashboards/logs_dashboard.json:111]: Same, hardcoded unique identifier?

File: scripts/deploy-frontend.sh
Comment (elipe17) [reply scripts/deploy-frontend.sh:16]: It will be needed when we migrate PLG to production for routing. I left it because I made PLG a public URL for the time being to expedite debugging and ease of access.

File: tdrs-backend/plg/deploy.sh
Comment (elipe17) [reply tdrs-backend/plg/deploy.sh:1]: It does not. It is intended to be ran manually from a local machine

File: tdrs-backend/plg/grafana/dashboards/logs_dashboard.json
Comment (elipe17) [reply tdrs-backend/plg/grafana/dashboards/logs_dashboard.json:28]: Yes, our datasources have UID's associated with them and we want our dashboards to reference those UIDs. If we don't do this, provisioning the dashboards on deploy/system start will not work.

File: tdrs-backend/plg/grafana/dashboards/logs_dashboard.json
Comment (elipe17) [reply tdrs-backend/plg/grafana/dashboards/logs_dashboard.json:111]: See above

File: tdrs-backend/docker-compose.yml
Comment (elipe17) [reply tdrs-backend/docker-compose.yml:99]: Not without running custom commands on the container to create the same structure. It is easier to use the directory structure the container already provides for us.

File: tdrs-backend/plg/prometheus/manifest.yml
Comment (elipe17) [reply tdrs-backend/plg/prometheus/manifest.yml:14]: [Grafana Mimir](https://github.com/grafana/mimir) supports remote read and write by way of S3. This is probably the only solution we can use in Cloud.gov to support long term metric storage.

Comment (elipe17) [reply tdrs-backend/plg/prometheus/manifest.yml:14]: [Prometheus Remote Storage Integrations](https://prometheus.io/docs/operating/integrations/)

File: tdrs-backend/plg/grafana/manifest.yml
Comment (andrew-jameson) [inline tdrs-backend/plg/grafana/manifest.yml:24]: Created #3232 for this.

**Comment by ADPennington** [comment]
@elipe17 can you remind me: is there a relationship between this PR and #3199? I see some file changes here that bump-up RAM resources, so I'd like to have OCIO approval before we increase costs if possible. [reference to our questions](https://github.com/raft-tech/TANF-app/pull/3210#pullrequestreview-2357142530)

**Comment by elipe17** [comment]
> @elipe17 can you remind me: is there a relationship between this PR and #3199? I see some file changes here that bump-up RAM resources, so I'd like to have OCIO approval before we increase costs if possible. [reference to our questions](https://github.com/raft-tech/TANF-app/pull/3210#pullrequestreview-2357142530)

@ADPennington, There is a relationship between this and the ADR. The ADR provides a brief "why" with respect to the introduction of PLG and where it lives in the boundary diagram. The ADR does not cover the cost with respect to RAM for PLG, that could definitely be added. There is not any increase in RAM allocation for existing apps (frontend/backend/etc). However, I did provide general "guestimates" of the RAM needed for PLG to run smoothly while monitoring all of our deployments. That number is an aggregate needed for PLG in it's entirety and it is not required per space or anything like that. As an absolute maximum we would be looking at 10GB of memory and 3.5GB on the lowest end. I am pretty confident we will skew towards the low end. Just need to do some maths and better estimations to figure it out.

**Comment by elipe17** [comment]
> @elipe17 can you remind me: is there a relationship between this PR and #3199? I see some file changes here that bump-up RAM resources, so I'd like to have OCIO approval before we increase costs if possible. [reference to our questions](https://github.com/raft-tech/TANF-app/pull/3210#pullrequestreview-2357142530)

@ADPennington I also wanted to add that merging this wont incur any new cost. Everything is already deployed in the dev space with the resources we currently have. The cost implications will come into question when [3222](https://github.com/raft-tech/TANF-app/issues/3222) is worked.

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Comment by elipe17** [comment]
Below are my true calculations for the required memory to deploy PLG in production.

Prometheus is scraping about 2400 series from our dev environment. You can determine this by logging into Grafana and querying the`prometheus_tsdb_head_series` metric. Some simple math says 2400 * 3 is about 7200 series in total for Prometheus to manage with our current configuration across all deployed environments. The dev team would like to (in the future) add other series to be tracked for other monitoring and alerting solutions (CPU, memory, Elastic usage, etc...). Looking ahead, it would make sense to pad Prometheus' memory in preparation for those potential additions. We should expect to have ~2 or 3 times as many series when those are added; to be safe I assume 3 times as many, or 21,600 series across all environments. Using the tool provided [here](https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion/) I calculated that Prometheus will need ~100MB of memory to manage all of our series and future series (see screenshot below). From there we add the memory usage that the dev deployed Prometheus app is currently using (see screenshot below) and see that the app will require a minimum of 180MB or memory. We should pad this value for safety and availability. Thus, the current configuration of Prometheus with 512MB of memory should be sufficient for our use cases.

Screenshot 2024-10-24 at 9 45 15 AM

Screenshot 2024-10-24 at 9 47 53 AM

For Loki, we don't have as nice of a tool and we have to do a little bit of estimation. On my local machine, at rest with no logging or log querying happening Loki consumes ~100MB of memory. When parsing a 50MB file with logging set to it's most verbose setting (the case in dev/staging), and Grafana querying for new logs every 5 seconds (as fast as it will go), Loki's memory consumption peaked at ~250MB. With that in mind, when Loki is receiving logs from all environments there is a possibility that it's memory could increase by a factor of at least 6 (for each backend app). The likelihood of that all backend apps are parsing files at the same time is very low. But there is a good chance one backend app from each environment could parse a hefty file at the same time. With that said, starting Loki with a memory usage of 1GB would be my recommendation. We should also keep in mind that there are use cases that cause [Loki memory consumption to grow without bounds](https://github.com/grafana/loki/issues/6501). I was able to produce this locally and is something for future consideration.

For Grafana, no matter what I did in the deployed or local environment it's consumption never breached 150MB. [Grafana's documentation](https://grafana.com/docs/grafana/latest/setup-grafana/installation/#hardware-recommendations) also suggests that the minimum memory should be set at 512MB which is how it is configured in this PR and is my recommended configuration.

Including the Promtail deployments and the Postgres exporters we should expect a initial total memory requirement for PLG to be 2.5GB:
512MB (Prometheus)
512MB (Grafana)
512MB (PG exporters & Promtails)
1GB (Loki)

Because all of these apps are stateless, we can always scale their memory up or down as we learn more and more about their needs with respect to our workloads while they are deployed.

cc. @ADPennington, @andrew-jameson

**Review by ADPennington — APPROVED** [review]
thanks for pairing on review @elipe17 🚀 

- RAM quota updated in org, so we can bump when needed
- uid is not a secret
- let's check the zap scan results next week (after this merges) to see if it catches anything related to header settings

**Pull request merged**
